### PR TITLE
[OTel] Refactor of psycopg2 instrumentation

### PR DESCRIPTION
**Pull request by CagriYonca**
- Coverage 96%

### CODE REVIEW

File: src/instana/instrumentation/psycopg2.py [hunk 1]
```diff
@@ -5,33 +5,44 @@
...
 
-    @wrapt.patch_function_wrapper('psycopg2.extensions', 'register_type')
-    def register_type_with_instana(wrapped, instance, args, kwargs):
+    @wrapt.patch_function_wrapper("psycopg2.extensions", "register_type")
+    def register_type_with_instana(
+        wrapped: Callable[..., Any],
+        instance: Optional[Any],
+        args: Tuple[Any, ...],
+        kwargs: Dict[str, Any],
+    ) -> Callable[..., Any]:
```
Comment (GSVarsha) [inline src/instana/instrumentation/psycopg2.py:28]: Can we be more specific about the return type of the callable? If its too complicated, you can change it to below.

```suggestion
        wrapped: Callable[..., object],
        instance: Optional[Any],
        args: Tuple[Any, ...],
        kwargs: Dict[str, Any],
    ) -> Callable[..., object]:
```

File: src/instana/instrumentation/psycopg2.py [hunk 2]
```diff
@@ -5,33 +5,44 @@
...
+    @wrapt.patch_function_wrapper("psycopg2._json", "register_json")
+    def register_json_with_instana(
+        wrapped: Callable[..., Any],
+        instance: Optional[Any],
+        args: Tuple[Any, ...],
+        kwargs: Dict[str, Any],
+    ) -> Callable[..., Any]:
+        if "conn_or_curs" in kwargs:
+            if hasattr(kwargs["conn_or_curs"], "__wrapped__"):
+                kwargs["conn_or_curs"] = kwargs["conn_or_curs"].__wrapped__
```
Comment (GSVarsha) [inline src/instana/instrumentation/psycopg2.py:45]: Same as [this](https://github.com/instana/python-sensor/pull/597/files#r1734075703) comment.

**Review by GSVarsha — CHANGES_REQUESTED** [review]
Can we include the tests for [pep 249](https://github.com/instana/python-sensor/pull/597/files#diff-e52e4ddd58b7ef887ab03c04116e676f6280b824ab7469d5d3080e5cba4f2128R91) now that `psycopg2` instrumentation is also going in?

File: tests/clients/test_psycopg2.py [hunk 3]
```diff
@@ -48,341 +50,366 @@ def setUp(self):
...
+        with tracer.start_as_current_span("test"):
             self.cursor.execute("""SELECT * from users""")
             affected_rows = self.cursor.rowcount
             result = self.cursor.fetchone()
             self.db.commit()
 
-        self.assertEqual(1, affected_rows)
-        self.assertEqual(6, len(result))
+        assert 1 == affected_rows
+        assert 6 == len(result)
```
Comment (GSVarsha) [inline tests/clients/test_psycopg2.py:92]: You may have to make similar changes at multiple places in this file.

```suggestion
        assert affected_rows == 1
        assert len(result) == 6
```

File: tests/clients/test_psycopg2.py [hunk 4]
```diff
@@ -48,341 +50,366 @@ def setUp(self):
...
         spans = self.recorder.queued_spans()
-        self.assertEqual(2, len(spans))
+        assert 2 == len(spans)
 
         db_span, test_span = spans
 
-        self.assertEqual("test", test_span.data["sdk"]["name"])
-        self.assertEqual(test_span.t, db_span.t)
-        self.assertEqual(db_span.p, test_span.s)
+        assert "test" == test_span.data["sdk"]["name"]
```
Comment (GSVarsha) [inline tests/clients/test_psycopg2.py:99]: Same as [this](https://github.com/instana/python-sensor/pull/597/files#r1734086763) comment.

File: tests/clients/test_psycopg2.py [hunk 5]
```diff
@@ -48,341 +50,366 @@ def setUp(self):
...
+        assert test_span.t == db_span.t
+        assert db_span.p == test_span.s
 
-        self.assertIsNone(db_span.ec)
+        assert db_span.ec is None
```
Comment (GSVarsha) [inline tests/clients/test_psycopg2.py:103]: You may have to make similar changes at multiple places in this file.

```suggestion
        assert not db_span.ec
```

Comment (CagriYonca) [reply tests/clients/test_psycopg2.py:103]: Done

File: tests/clients/test_psycopg2.py (hunk 4 above)
Comment (CagriYonca) [reply tests/clients/test_psycopg2.py:99]: Done

File: tests/clients/test_psycopg2.py (hunk 3 above)
Comment (CagriYonca) [reply tests/clients/test_psycopg2.py:92]: Done

File: src/instana/instrumentation/psycopg2.py (hunk 2 above)
Comment (CagriYonca) [reply src/instana/instrumentation/psycopg2.py:45]: Done

File: src/instana/instrumentation/psycopg2.py (hunk 1 above)
Comment (CagriYonca) [reply src/instana/instrumentation/psycopg2.py:28]: Done

**Review by pvital — APPROVED** [review]
It looks good to me.

**Comment by pvital** [comment]
I think you should rebase first since I see commits that are already on `otel_migration` branch.

**Pull request merged**